dogma.go

```diff
@@ -262,42 +266,84 @@ type errorMessage struct {
 	Message string `json:"message"`
 }
 
-func (c *Client) do(ctx context.Context, req *http.Request, resContent interface{}) (int, error) {
-	req = req.WithContext(ctx)
-	res, err := c.client.Do(req)
-	if err != nil {
-		return UnknownHttpStatusCode, err
-	}
-	defer func() {
+func drainupAndCloseResponseBody(body io.ReadCloser) {
+	if body != nil {
 		// drain up 512 bytes and close the body to reuse connection
 		// see also:
 		// - https://github.com/google/go-github/pull/317
 		// - https://forum.golangbridge.org/t/do-i-need-to-read-the-body-before-close-it/5594/4
-		io.CopyN(ioutil.Discard, res.Body, 512)
+		io.CopyN(ioutil.Discard, body, 512)
+
+		// close body
+		body.Close()
+	}
+}
 
-		res.Body.Close()
-	}()
+func (c *Client) do(ctx context.Context, req *http.Request, resContent interface{}) (statusCode int, err error) {
+	req = req.WithContext(ctx)
+
+	// prepare metrics
+	var metricLabels []metrics.Label
+	if c.metricCollector != nil {
+		metricLabels = []metrics.Label{
+			{Name: "method", Value: req.Method},
+			{Name: "host", Value: req.URL.Host},              // included port
+			{Name: "path", Value: req.URL.EscapedPath()},     // escaped path
+			{Name: "query", Value: req.URL.Query().Encode()}, // encoded query
+		}
+	}
+
+	// mark the time point when request begins
+	startAt := time.Now()
+
+	// make request
+	res, err := c.client.Do(req)
+
+	// get response status code
+	if err == nil {
+		statusCode = res.StatusCode
+	} else {
+		statusCode = UnknownHttpStatusCode
+	}
 
-	if res.StatusCode < 200 || res.StatusCode >= 300 {
+	// report metric
+	if c.metricCollector != nil {
+		metricLabels = append(metricLabels, metrics.Label{Name: "statusCode", Value: strconv.Itoa(statusCode)})
+		c.metricCollector.MeasureSinceWithLabels([]string{"requestDuration"}, startAt, metricLabels)
+	}
+
+	// check request error
+	if err != nil {
+		return
```

Comment:
I'm curious if we need to count the request failure?

Reply:
🤔 Sound reasonable. Let me make a commit soon.